chain/store/messages.go

```diff
@@ -300,3 +301,144 @@ func (cs *ChainStore) LoadSignedMessagesFromCids(ctx context.Context, cids []cid
 
 	return msgs, nil
 }
+
+// TipSetBlockMessagesReceipts returns the blocks and messages in `ts` and their corresponding receipts from `pts` matching block order in tipset (`ts`).
+func (cs *ChainStore) TipSetBlockMessagesReceipts(ctx context.Context, ts, pts *types.TipSet) ([]*BlockMessageReceipts, error) {
+	// returned BlockMessages match block order in tipset
+	blkMsgs, err := cs.BlockMsgsForTipset(ctx, pts)
+	if err != nil {
+		return nil, err
+	}
+	if len(blkMsgs) != len(pts.Blocks()) {
+		// logic error somewhere
+		return nil, fmt.Errorf("mismatching number of blocks returned from block messages, got %d wanted %d", len(blkMsgs), len(pts.Blocks()))
+	}
+
+	// retrieve receipts using a block from the child (ts) tipset
+	// TODO this operation can fail when the node is imported from a snapshot that does not contain receipts (most don't)
+	// the solution is to compute the tipset state which will create the receipts we load here
+	rs, err := blockadt.AsArray(cs.ActorStore(ctx), ts.Blocks()[0].ParentMessageReceipts)
```

Comment:
I'd add a `ParentMessageReceipts` method to the `TipSet` type for this

Reply:
done via [a07ee7e](https://github.com/filecoin-project/lotus/pull/9186/commits/a07ee7e6ef3d5b124efcf5213b76f598649c6281)